### ISSUE TITLE
Can't get baggage to propagate with boot 3 and micrometer tracing

### ISSUE THREAD

**Issue by eitansuez** [micrometer-metrics/micrometer#3564]
**Describe the bug**
I can't get remote baggage fields to propagate using spring boot 3 and micrometer tracing. 

**Environment**
Local machine
 - Micrometer version: 1.0.0
 - Micrometer registry : N/A
 - OS: macOS
 - Java version: 19.0.1

**To Reproduce**
How to reproduce the bug:
1. See project: https://github.com/eitansuez/dtrace-boot3-example which contains an automated test that fails, or..
2. ..Use the micrometer sample module "baggage consumer" https://github.com/micrometer-metrics/micrometer-samples/tree/main/baggage-consumer as follows: start the app (gradle bootRun), then curl the endpoint at localhost:7200/ with the header "myremotefield" and value "my-remote-field-value" and note how the RemoteFieldsTestChecker returns an error.

**Expected behavior**
If i define a baggage remote-field to propagate, i expect that field to show up in http requests to upstream services.

**Additional context**
Slack thread: https://micrometer-metrics.slack.com/archives/C662HUJC9/p1670600358434289

**Comment by marcingrzejszczak** [comment]
I think this might be a duplicate of https://github.com/micrometer-metrics/tracing/issues/122

**Comment by akoufa** [comment]
@marcingrzejszczak Do you know when version 1.0.1 will be released ? I saw the milestone is already complete. Is there any snapshot release available ?

**Comment by marcingrzejszczak** [comment]
Yes, it's available. Just add `https://repo.spring.io/snapshot` as a maven dependency.

We haven't yet decided when we will do a release cc @shakuzen @jonatan-ivanov

**Comment by akoufa** [comment]
@marcingrzejszczak  Strange, I don't know if I have defined something wrongly but it seems Maven cannot resolve the 1.0.1 snapshot release. 

```
<dependency>
      <groupId>io.micrometer</groupId>
      <artifactId>micrometer-tracing-bridge-otel</artifactId>
      <version>1.0.1-SNAPSHOT</version>
    </dependency>
```

**Comment by akoufa** [comment]
@marcingrzejszczak Unfortunately, it does not seem to work with 1.0.1 . `Slf4JBaggageEventListener` is being called but baggage is still empty

Also manually creating baggage with the `createBaggage` method seems to always set the value of the baggage entry to `""` an empty string. See the following line of code:

https://github.com/micrometer-metrics/tracing/blob/03576ff2cb57463147e47a6fc2cc46b93cc94e25/micrometer-tracing-bridges/micrometer-tracing-bridge-otel/src/main/java/io/micrometer/tracing/otel/bridge/OtelBaggageManager.java#L147

Is it expected that `baggageWithValue(name, "")` is called with an empty string despite passing some value in the outer method?

**Comment by marcingrzejszczak** [comment]
Have you actually put things in scope? How do you use the baggage? I mean we have tests and samples that show that baggage is indeed working

**Comment by eitansuez** [comment]
the micrometer-samples indeed has an acceptance test that runs the baggage-consumer app, then runs a second app, baggage-producer that sends an http request to the consumer.  the test verifies that baggage is indeed propagated.  however, a variant test: where the initial request is made with `curl` shows that baggage does not propagate.  the difference as far as i can tell is the presence of some kind of "traceparent" header (in the former case).

what led me down this path was an attempt to create a variant of the [spring petclinic for istio](https://github.com/eitansuez/spring-petclinic-istio).  i stood up the application, made some token requests and would observe in jaeger that all my traces had only two spans, a clear indication that [the needed tracer headers](https://istio.io/latest/docs/tasks/observability/distributed-tracing/overview/) weren't being propagated.

**Comment by marcingrzejszczak** [comment]
Can we agree that this https://github.com/spring-projects/spring-boot/issues/33516#issuecomment-1351714194 issue duplicates this one? Then we could close the one in Boot and fix any issues here.

From what I see the samples are working fine because the baggage gets added to when rest template call is made. Rest template is instrumented so we're sending the full tracing info and the baggage in addition to that. When you're making the call with curl, the tracing headers are missing and we're not supporting that at this point (which seems like a bug).

Are we in agreement with the findings so far?

**Comment by eitansuez** [comment]
Agreed.  They are both the same issue as far as i can tell.

**Comment by marcingrzejszczak** [comment]
Actually I found your project (https://github.com/eitansuez/dtrace-boot3-example). Add `@AutoConfigureObservability` to your test and it will work fine. Otherwise you're using the `SimpleTracer` and that will not work cause it doesn't know how to resolve baggage properly from headers

**Comment by eitansuez** [comment]
indeed, adding the annotation to the test makes it pass.  but when i `gradle bootRun` the app and `curl` the endpoint, i still don't see the propagation happening.  i'm not entirely sure that my build is properly configured with version 1.0.1-SNAPSHOT of all micrometer dependencies.  i pushed my `build.gradle` changes so you can review/make sure i have that configured correctly.

**Comment by marcingrzejszczak** [comment]
Yup, good catch. The `RestTemplate` in tests was instrumented it seems. I did do another fix (https://github.com/micrometer-metrics/tracing/commit/fc08acca14701355f1ab0d22b9561e9dcca2cc1f).

**Comment by marcingrzejszczak** [comment]
@eitansuez can you confirm that the issue got fixed?

**Comment by eitansuez** [comment]
the way i test:

```shell
git clone https://github.com/eitansuez/dtrace-boot3-example
cd dtrace-boot3-example
gradle bootRun
```

and in a separate terminal:

```shell
curl localhost:8080 -H "x-request-id: world"
```

and i look for a response body matching: "Hello, world"

currently, i'm still getting "Hello, null"

the build dependencies use version 1.0.1-SNAPSHOT of the micrometer tracing bom.

the header `x-request-id` is specified in `application.yaml` as a baggage remote field.  that endpoint makes a call to another endpoint.  that second endpoint returns string body containing the contents of the x-request-id header in the propagated request.

**Comment by eitansuez** [comment]
i should mention that this very same scenario works with the latest stable 2.x version of spring boot + spring cloud sleuth.

**Comment by marcingrzejszczak** [comment]
This is my gradle build

```groovy
plugins {
  id 'java'
  id 'org.springframework.boot' version '3.0.0'
  id 'io.spring.dependency-management' version '1.1.0'
}

group = 'com.example'
sourceCompatibility = '19'

repositories {
  mavenLocal()
  mavenCentral()
}

dependencies {
  implementation 'org.springframework.boot:spring-boot-starter-web'
  implementation 'org.springframework.boot:spring-boot-starter-actuator'
  implementation 'io.micrometer:micrometer-tracing-bridge-otel:1.0.1-SNAPSHOT'

  testImplementation 'org.springframework.boot:spring-boot-starter-test'
}

tasks.named('test') {
  useJUnitPlatform()
}

```

This is my curl

```
curl localhost:8080 -H "x-request-id: world"
```

This is my result

```
➜  ~ curl localhost:8080 -H "x-request-id: world"
Hello, world%  
```

I guess you're fetching some old snapshots?

**Comment by eitansuez** [comment]
it's possible that my local snapshots are stale.
the main difference i see is that i'm referencing the brave bridge.  happy to switch to otel.
i can confirm it works on my end too using the otel bridge.

**Comment by marcingrzejszczak** [comment]
I'm just about to apply a fix for Brave

**Comment by eitansuez** [comment]
super.  thank you!!

**Comment by marcingrzejszczak** [comment]
Fixed via https://github.com/micrometer-metrics/tracing/issues/131, https://github.com/micrometer-metrics/tracing/issues/130 and https://github.com/micrometer-metrics/tracing/issues/129

**Comment by marcingrzejszczak** [comment]
BTW thanks for a great reproducer!

**Comment by eitansuez** [comment]
it's my pleasure.  thanks!

**Comment by kishorepurushothaman20** [comment]
Hi,

Have a small problem in header propagation to upstream services,
1. The propagated headers are always translated to smaller case to upstream service. Example `SSLClientId` is always `sslclientid` when call goes to upstream service
2. When using spring webclient and subscribing the reactive stream with a scheduler like `Schedulers.parallel()`, the context is lost meaning Baggage remote fields are not propagated as well as new traceId and spanId is created.
Example,
`webClient
                .post()
                .uri(URL)
                .bodyValue(request)
                .retrieve()
                .subscribeOn(Schedulers.parallel())
`

**Comment by marcingrzejszczak** [comment]
Wait for the next release of Spring Boot that includes a lot of fixes for reactive and try again and if that still doesn't work then comment on this issue again please

**Comment by mikemybytes** [comment]
To whom it may concern: when using Reactor (e.g. Spring WebFlux), you probably have to enable _automatic context propagation_ to make this work:
```java
// register once, e.g. in the app's main method
Hooks.enableAutomaticContextPropagation();
```
See [Reactor Hooks docs](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Hooks.html#enableAutomaticContextPropagation--) and the [explanation on Spring blog](https://spring.io/blog/2023/03/30/context-propagation-with-project-reactor-3-unified-bridging-between-reactive).

**Comment by ShaneWhelan** [comment]
@mikemybytes Thanks for the comment around getting this to work for WebFlux!

That solved some of my problems but I also had problems with message brokers like Kafka and Google Pub/Sub. Since the migration from Sleuth to Micrometer, I added `io.micrometer:micrometer-tracing-bridge-brave` but I also needed to add `io.zipkin.brave:brave-instrumentation-messaging` to enable tracing with a message broker.

For integration tests the `@AutoConfigureObservability` annotation needs to be added to tests to enable observability with Kafka and Google Pub/Sub.

Also for anyone who is trying to get this working for Kafka, you also need this `setObservationEnabled` setting on your Producer config, here is one approach to achieve that:
``` java
  @Bean
  public KafkaTemplate<String, SomeAvroObject> kafkaTemplate(ProducerFactory<String, SomeAvroObject> producerFactory) {
    var kafkaTemplate = new KafkaTemplate<>(producerFactory);
    // Set Kafka to enable micrometer tracing and header propagation
    kafkaTemplate.setObservationEnabled(true);
    return kafkaTemplate;
  }
```
In Consumer config you also need `setObservationEnabled`, here is one approach to achieve that:
``` java
  @Bean
  public ConcurrentKafkaListenerContainerFactory<String, String> kafkaListenerContainerFactory() {
    ConcurrentKafkaListenerContainerFactory<String, String> factory = new ConcurrentKafkaListenerContainerFactory<>();
    factory.setConsumerFactory(consumerFactory());
    factory.getContainerProperties().setObservationEnabled(true);
    return factory;
  }
```

For Google Pub/Sub you still needed `spring.cloud.gcp.trace.pubsub.enabled=true` with Sleuth, but its also needed for Micrometer so ensure thats set somewhere.

Now any baggage will propagate over outbound HTTP calls and when a message is produced to message broker or consumed from one.

**Comment by jonatan-ivanov** [comment]
Fyi: Spring Kakfa is instrumented, you should be able to enable it instead of adding `brave-instrumentation-messaging`.

**Comment by ShaneWhelan** [comment]
> Fyi: Spring Kakfa is instrumented, you should be able to enable it instead of adding `brave-instrumentation-messaging`.

Thanks! I ran a test to check what I actually need for Kafka and I don't need `brave-instrumentation-messaging`, but I did need `setObservationEnabled` on the Producer and Consumer (which was not needed before on Sleuth).

**Comment by marcingrzejszczak** [comment]
I think they write about this in [their docs](https://docs.spring.io/spring-kafka/reference/kafka/micrometer.html#observation)

**Comment by zglueck** [comment]
Thanks @mikemybytes and @ShaneWhelan this unblocked by baggage propagation woes. And @ShaneWhelan sorry I didn't get a chance to say goodbye, funny I should run into your help once again here. 👍
